lxd/storage_volumes_config.go

```diff
@@ -0,0 +1,125 @@
+package main
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/lxc/lxd/shared"
+	"github.com/lxc/lxd/shared/api"
+)
+
+var storageVolumeConfigKeys = map[string]func(value string) error{
+	"block.mount_options":  shared.IsAny,
+	"block.filesystem":     shared.IsAny,
+	"size":                 shared.IsAny,
+	"zfs.use_refquota":     shared.IsBool,
+	"zfs.remove_snapshots": shared.IsBool,
+}
+
+func storageVolumeValidateConfig(name string, config map[string]string, parentPool *api.StoragePool) error {
+	for key, val := range config {
+		// User keys are not validated.
+		if strings.HasPrefix(key, "user.") {
+			continue
+		}
+
+		// Validate storage volume config keys.
+		validator, ok := storageVolumeConfigKeys[key]
+		if !ok {
+			return fmt.Errorf("Invalid storage volume configuration key: %s", key)
+		}
+
+		err := validator(val)
+		if err != nil {
+			return err
+		}
+
+		if parentPool.PoolDriver != "zfs" || parentPool.PoolDriver == "dir" {
+			if config["zfs.use_refquota"] != "" {
+				return fmt.Errorf("Key volume.zfs.use_refquota cannot be used with non zfs storage volumes.")
+			}
+
+			if config["zfs.remove_snapshots"] != "" {
+				return fmt.Errorf("Key volume.zfs.remove_snapshots cannot be used with non zfs storage volumes.")
+			}
+		}
+
+		if parentPool.PoolDriver == "dir" {
+			if config["block.mount_options"] != "" {
+				return fmt.Errorf("Key block.mount_options cannot be used with dir storage volumes.")
+			}
+
+			if config["block.filesystem"] != "" {
+				return fmt.Errorf("Key block.filesystem cannot be used with dir storage volumes.")
+			}
+
+			if config["size"] != "" {
+				return fmt.Errorf("Key size cannot be used with dir storage volumes.")
+			}
```

Comment:
Same problem as with the wording of the pool errors, this should be changed for:

"The \"ABC\" key cannot be used with ..."